### PR TITLE
machines: replace ToastNotificationList with AlertGroup PF4 component

### PR DESCRIPTION
**Pull request by KKoukiou**
Toast Notifications how they were before they were a result of a lot of custom code, PF4 Toast notifications are just a plain ul list on the right of the page.

If we really need the custom things, we should also post a patch in PF.

Disclaimer: This is not urgent

### CODE REVIEW

**Comment by KKoukiou** [comment]
Screen Shot 2020-10-12 at 17 00 10


Without any custom CSS.

**Comment by KKoukiou** [comment]
@garrett, @andreasn do you think we should block replacement of this component on having a local override CSS as we did before?

**Comment by garrett** [comment]
The goal of the local CSS overrides were to prevent errors from getting in the way of the VMs. Sometimes, several notifications would occur at once and then it wasn't even possible to see which VM was affected. And they were initially very verbose by default.

Original reasoning: https://github.com/cockpit-project/cockpit/pull/11419#issuecomment-474930638
Why they were in the center: https://github.com/cockpit-project/cockpit/pull/11419#issuecomment-475180374

The main goal was to handle multiple notifications at once and to have them all out of the way of any critical content, which includes:
- Create & Import VM buttons (if the error was during creation, couldn't immediately try again)
- VM statuses (couldn't even see which VM was having problems in some cases, if it was at the top)
- VM actions (start/stop/etc.)

Secondary goals were:
- anchor notifications to the area where they were related to; that is: make placement relative the VM list, not the page
- fix additional issues that showed up in smaller screen sizes (including, but not limited, to mobile) — most notably: the notifications would be relatively larger, including wrapping, so they covered everything on the screen

**Comment by garrett** [comment]
Also:

- It's great that the additional text is collapsed with a "show more" link by default already.
- Even if we did ported the overrides over, it would hopefully be simpler this time around.
- It's odd that icon alignment is not right. Is that depending on CSS that isn't applied or something?

I'm currently thinking if there's some alternate way that might work to show these.

PF's notification drawer would require shell modifications and force a user to do notification management and it would also block the content even more. These are both things that we want to try to avoid doing.

**Comment by garrett** [comment]
I wonder if we could make the errors be more contextual.

1. Creation errors could be an inline notification to the side of (or even below or above?) the create buttons.
2. VM-specific errors could be just the most recent error (if > 1 error) on the error state with a popover for more information.

Doing this, I think we might not have to do anything custom, including notification stacking.

**Comment by garrett** [comment]
I was thinking something like this:

vm-list-err

The popover is just a placeholder to show that it'd have a popover. In the implementation, it'd be larger and such.

I've used labels for the status. (That's the only [error state for tables/lists](https://www.patternfly.org/v4/components/table/design-guidelines/#default-spacing) I can find in any PF4 examples.)

The shut off state is a label too — but, to demphasize it, I think we'd want a gray one and just hide the background (in CSS). That way, it'd use the same code and the text would be the same size and line up with the rest of the statuses.

**Comment by garrett** [comment]
The mockup above uses PF4 widgets with only a 1-line override for the shut off state (as mentioned above).

For create/import errors, we could either do:

1. inline alert near the create/import buttons (as mentioned in a prior comment of mine)
2. alert notification, but repositioned as to not overlap the create/import buttons

Since the other errors would be inline with the VMs, we wouldn't need to worry about stacking or any other complex stuff.

**Comment by garrett** [comment]
Discussed this in IRC with @KKoukiou: To make the notifications dismissable, we'll use the × icon on the right, which is a standard part of PF4 labels.

Also, if there's an error for the VM (like it couldn't start for some reason) and then runs successfully, it should automatically clear the error. (Not sure if that's the case already.)

**Comment by KKoukiou** [comment]
@garrett another thought that I got when looking at the mockup:

When a VM had an operation fail it's state is not failed. It can be 'Shut off' because it failed to Start for example.
Or 'Paused' because it failed to resume. I wonder if you expect that the UI will disable the correct state with the red label there, like the screenshot below:
Screen Shot 2020-10-16 at 09 45 21
Or you expect to have the 'Failed' label whetever the actual state is.

**Comment by garrett** [comment]
The state would be separate items.

(some error) (running)
(some error) (shut off)

**Comment by KKoukiou** [comment]
> The state would be separate items.
> 
> (some error) (running)
> (some error) (shut off)

So you would like a state like:

'[Failed show more..] Running'

**Comment by garrett** [comment]
Correct.

**Comment by KKoukiou** [comment]
@garrett it currently looks like this:
Screen Shot 2020-10-27 at 17 12 52

**Comment by garrett** [comment]
I wonder if we should move the general status first, like shut off, running, etc. and _then_ the error, instead of having the error first.

**Review by garrett — DISMISSED** [review]
I wonder if we should move the general status first, like shut off, running, etc. and then the error, instead of having the error first.

Otherwise, I think this is pretty much ready to merge. Looks great!

File: pkg/machines/components/stateIcon.jsx [hunk 1]
```diff
@@ -62,7 +62,7 @@ export const StateIcon = ({ state, valueId, error, dismissError }) => {
             </Label>}
             <Label color={stateMap[state] && stateMap[state].color}
                    icon={stateMap[state] && stateMap[state].icon}
-                   className="vm-state-text"
+                   className={"vm-state-text vm-state--" + state.toLowerCase().replaceAll(' ', '-')}
```
Comment (garrett) [inline pkg/machines/components/stateIcon.jsx:65]: Is there a better way to do this, or is this fine?

Comment (garrett) [reply pkg/machines/components/stateIcon.jsx:65]: (The goal is to have a class name to attach some CSS to, so we can de-emphasize a shut off state.)

**Comment by KKoukiou** [comment]
@garrett With the CSS custom code for labels in the tables.

Screen Shot 2020-10-28 at 14 19 32

Without any custom CSS for labels in the tables.
Screen Shot 2020-10-28 at 14 19 17

It seems to me super close, why do we need it?

File: pkg/machines/components/stateIcon.scss [hunk 2]
```diff
@@ -3,12 +3,16 @@
 
   // In definition lists (for Overview),
   // match text size and alignment
-  .pf-c-description-list__description & {
+  .pf-c-table &,
```
Comment (KKoukiou) [inline pkg/machines/components/stateIcon.scss:6]: @garrett With the CSS custom code for labels in the tables.

Screen Shot 2020-10-28 at 14 19 32

Without any custom CSS for labels in the tables.
Screen Shot 2020-10-28 at 14 19 17

It seems to me super close, why do we need it?

Comment (KKoukiou) [reply pkg/machines/components/stateIcon.scss:6]: Ok I actually see why we need the margin-left: calc(var(--pf-c-label--PaddingLeft)*-1);

But all the rest?

Comment (garrett) [reply pkg/machines/components/stateIcon.scss:6]: 1. In the list, as seen above, it's not completely necessary (but doesn't hurt)
2. In the mobile version of the list, things do not align without it
3. In the details page, things do not align without it

**Comment by KKoukiou** [comment]
Ah we forget the custom CSS for the storage and networks lists.
I generalized it.

File: pkg/machines/components/stateIcon.jsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                className="resource-state-text"
+                onClose={dismissError}
+                id={valueId}>
+                <>
+                    {_("Failed")}
+                    <Button variant="link" isInline>{_("view more...")}</Button>
+                </>
+            </Label>}
+            <Label color={stateMap[state] && stateMap[state].color}
+                   icon={stateMap[state] && stateMap[state].icon}
```
Comment (marusak) [inline pkg/machines/components/stateIcon.jsx:64]: `icon` is never defined. I guess because you defined icons as classNames?

**Review by marusak — CHANGES_REQUESTED** [review]
Super cool!
I could not get any error on storage pools page. But I assume it works :) If you can help me with 
reproducer or just post screenshot that would be great :)

File: pkg/machines/components/stateIcon.jsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2016 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (garrett) [reply pkg/machines/components/stateIcon.jsx:4]: In general: Copyright strings don't really need to be updated. (I don't know if there's any policy for RH and code or such.)

[the remark this answers is not in the document]
Comment (garrett) [reply pkg/machines/components/stateIcon.jsx:4]: Some info on it: https://techwhirl.com/updating-copyright-notices/

Also, having "copyright copyright" is redundant. (The (C) is standing in for ©, which literally means copyright.)

File: pkg/machines/components/stateIcon.jsx (hunk 3 above)
Comment (KKoukiou) [reply pkg/machines/components/stateIcon.jsx:64]: Right, it was meant to be icon above.
Therefore the icons were missing from there UI.
Currently it looks like:
Screen Shot 2020-11-03 at 15 52 16

**Comment by KKoukiou** [comment]
> Super cool!
> I could not get any error on storage pools page. But I assume it works :) If you can help me with
> reproducer or just post screenshot that would be great :)

I had an actually bug in the code, and storage pools errors were not updated. Now they are and they look like this:

Screen Shot 2020-11-03 at 15 48 34

**Comment by marusak** [comment]
> I had an actually bug in the code, and storage pools errors were not updated. Now they are and they look like this:

Nice, thanks! I had one pool that I could not start but could not see any error so was not sure :)

**Review by marusak — APPROVED** [review]
Thanks, looks good now!
Tests are somewhat unhappy, but looks unrelated. But will retry.

**Comment by garrett** [comment]
The popovers are great. I thought it could use a little bit of alert style, however, so I pulled in some styles of an alert in commit 
4e88c96.

If you like this, we can go with it. We might want to even consider making a hybrid widget for Cockpit and/or pushing it upstream too. If you think it's too much for this PR, we can revert the change and/or move it to its own PR (and possibly push it into lib if it makes sense).

**Comment by KKoukiou** [comment]
@marusak can you please reapprove. garrett posted some CSS bits and dismissed your review

**Review by marusak — APPROVED** [review]

**Pull request merged**